server/mdm/nanomdm/service/nanomdm/service.go

```diff
@@ -253,25 +262,40 @@ func (s *Service) CommandAndReportResults(r *mdm.Request, results *mdm.CommandRe
 	if err != nil {
 		return nil, fmt.Errorf("retrieving next command: %w", err)
 	}
-	if cmd != nil {
+	if cmd == nil {
 		logger.Debug(
-			"msg", "command retrieved",
-			"command_uuid", cmd.CommandUUID,
-			"request_type", cmd.Command.RequestType,
+			"msg", "no command retrieved",
 		)
-		// We expand secrets in the command before returning it to the caller so that we never store unencrypted secrets in the database.
-		expanded, err := s.store.ExpandEmbeddedSecrets(r.Context, string(cmd.Raw))
-		if err != nil {
-			// This error is not expected since secrets should have been validated on profile upload.
-			logger.Info("level", "error", "msg", "expanding embedded secrets", "err", err)
-			// Since this error should not happen, we use the command as is, without expanding secrets.
-		} else {
-			cmd.Raw = []byte(expanded)
-		}
-		return cmd, nil
+		return nil, nil
 	}
 	logger.Debug(
-		"msg", "no command retrieved",
+		"msg", "command retrieved",
+		"command_uuid", cmd.CommandUUID,
+		"request_type", cmd.Command.Command.RequestType,
+		"subtype", cmd.Subtype,
 	)
-	return nil, nil
+	// We expand secrets in the command before returning it to the caller so that we never store unencrypted secrets in the database.
+	// User can issue a one-off MDM command that contains a secret variable, for example.
+	expanded, err := s.store.ExpandEmbeddedSecrets(r.Context, string(cmd.Raw))
+	if err != nil {
+		// This error should never happen since secrets have been validated on profile upload.
+		logger.Info("level", "error", "msg", "expanding embedded secrets", "err", err)
+		// Since this error should not happen, we simply use the command as is, without expanding secrets.
+	} else {
+		cmd.Raw = []byte(expanded)
+	}
+	switch cmd.Subtype {
+	case mdm.CommandSubtypeProfileWithSecrets:
+		// Secrets were expanded above. Now we need to base64 encode and sign the configuration profile before returning it to the caller.
```

Comment:
Have we looked into encrypting the profile content here? Otherwise, I'm assuming we are ok for this iteration that unencrypted secrets might be accessible in transit or at rest on the device, for example, by a user running `profiles` in the terminal or via osquery `macos_profiles` table.

Reply:
Yes, I brought up a similar issue with product.

Are you suggesting there is a way to prevent the device user from viewing profile contents. For example, from viewing the EnrollSecret in our Fleetd configuration profile?

Reply:
Yes, it is possible to encrypt profiles. Here are the Apple [docs](https://developer.apple.com/documentation/devicemanagement/configuring_multiple_devices_using_profiles#3234104).